ciri/core.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <string.h>
+
+#include "ciri/core.h"
+#include "utils/logger_helper.h"
+
+// TODO(thibault) from configuration variables
+#define CIRI_DB_PATH "ciri/ciri.db"
+#define CORE_LOGGER_ID "core"
+
+retcode_t core_init(core_t* const core) {
+  if (core == NULL) {
+    return RC_CORE_NULL_CORE;
+  }
+
+  logger_helper_init(CORE_LOGGER_ID, LOGGER_DEBUG, true);
+  memset(core, 0, sizeof(core_t));
+
+  log_info(CORE_LOGGER_ID, "Initializing database connection\n");
+  core->db_conf.db_path = CIRI_DB_PATH;
+  core->db_conf.index_address = true;
+  core->db_conf.index_approvee = true;
+  core->db_conf.index_bundle = true;
+  core->db_conf.index_tag = true;
+  if (init_connection(&core->db_conn, &core->db_conf)) {
+    log_critical(CORE_LOGGER_ID, "Initializing database connection failed\n");
+    return RC_CORE_FAILED_DATABASE_INIT;
+  }
+
+  log_info(CORE_LOGGER_ID, "Initializing cIRI node\n");
+  if (node_init(&core->node, core)) {
+    log_critical(CORE_LOGGER_ID, "Initializing cIRI node failed\n");
+    return RC_CORE_FAILED_NODE_INIT;
+  }
+  return RC_OK;
+}
+
+retcode_t core_start(core_t* const core) {
+  if (core == NULL) {
+    return RC_CORE_NULL_CORE;
+  }
+
+  log_info(CORE_LOGGER_ID, "Starting cIRI node\n");
+  if (node_start(&core->node)) {
+    log_critical(CORE_LOGGER_ID, "Starting cIRI node failed\n");
+    return RC_CORE_FAILED_NODE_START;
+  }
+  return RC_OK;
+}
+
+retcode_t core_stop(core_t* const core) {
+  retcode_t ret = RC_OK;
+
+  if (core == NULL) {
+    return RC_CORE_NULL_CORE;
+  }
+
+  log_info(CORE_LOGGER_ID, "Stopping cIRI node\n");
+  if (node_stop(&core->node)) {
+    log_error(CORE_LOGGER_ID, "Stopping cIRI node failed\n");
+    ret = RC_CORE_FAILED_NODE_STOP;
+  }
+  return ret;
+}
+
+retcode_t core_destroy(core_t* const core) {
+  retcode_t ret = RC_OK;
+
```

Comment:
How do you guarantee that core is stopped before it's destroyed? I guess that might also apply to other components

Reply:
As discussed, the field `running`is now checked before destroying a component.
It has also been added to `node` and `core`.